Bluetooth: controller: Add DF antenna configuration handling

ppryga-nordic opened the pull request:
PR adds:
- antennae configuration by use of DTS (including new binding description),
- verification of antennae configuration
- initialization of Radio settings according to conf. provided by DTS
- add interface between ULL and LLL to retrieve information about available antennae number.

Radio configuration is part of controller initialization sequence.

ppryga-nordic committed:
dts: arm: nordic: Add radio peripheral with DFE antenna config

Add radio peripheral to nrf52833 DTS including antenna matrix
congiuration for Direction Finding extennsion.
Add appropriate binding file for nRF radio peripheral.

There is no default antenna matrix configuration. Antennas number
and GPIOS mapping to DFEGPIOS is project specific.
Complete configuration must be provided by end user as overaly.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: controller: ll_sw: nordic: Add DF antenna configuration handling

Add support for Direction Finding antenna configuration in
Nordic hal of Bluetooth controller.
Add handling of response for HCI_LE_Read_Antenna_Information
HCI command in controller (upper and lower link layer).
Handling of response to the command requires antenna matrix
configuration.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: controller: Add DF initialization to controller init sequence

Direction Finding requires initalization of a set of registers
in Radio peripheral. Also it requires information about antenna
matrix unit that are provided by end user in DTS overlay.
Implemented initialization step is responsible for validation of
hardware information delivered by DTS and setting up DF related
registers in Radio.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_df.c:
```diff
@@ -60,11 +61,20 @@ void ll_df_read_ant_inf(uint8_t *switch_sample_rates,
-	*num_ant = 0;
-	*max_switch_pattern_len = 0;
-	*max_cte_len = 0;
+#if IS_ENABLED(CONFIG_BT_CTLR_DF_ANT_SWITCH_TX) && \
+	IS_ENABLED(CONFIG_BT_CTLR_DF_ANT_SWITCH_1US)
+	*switch_sample_rates |= DF_AOD_1US_TX;
+#endif
+#if IS_ENABLED(CONFIG_BT_CTLR_DF_CTE_RX) && \
+	IS_ENABLED(CONFIG_BT_CTLR_DF_CTE_RX_SAMPLE_1US)
+	*switch_sample_rates |= DF_AOD_1US_RX;
+#endif
+#if IS_ENABLED(CONFIG_BT_CTLR_DF_ANT_SWITCH_RX) && \
+	IS_ENABLED(CONFIG_BT_CTLR_DF_CTE_RX_SAMPLE_1US)
+	*switch_sample_rates |= DF_AOA_1US;
+#endif
```
```suggestion
	if (IS_ENABLED(CONFIG_BT_CTLR_DF_ANT_SWITCH_TX) &&
	    IS_ENABLED(CONFIG_BT_CTLR_DF_ANT_SWITCH_1US)) {
		*switch_sample_rates |= DF_AOD_1US_TX;
	}
	if (IS_ENABLED(CONFIG_BT_CTLR_DF_CTE_RX) &&
	    IS_ENABLED(CONFIG_BT_CTLR_DF_CTE_RX_SAMPLE_1US)) {
		*switch_sample_rates |= DF_AOD_1US_RX;
	}
	if (IS_ENABLED(CONFIG_BT_CTLR_DF_ANT_SWITCH_RX) &&
	    IS_ENABLED(CONFIG_BT_CTLR_DF_CTE_RX_SAMPLE_1US)) {
		*switch_sample_rates |= DF_AOA_1US;
	}
```

ppryga replied:
Changed to requested form.

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull.c:
```diff
@@ -415,6 +416,13 @@ int ll_init(struct k_sem *sem_rx)
 		ull_filter_reset(true);
 	}
 
+#if IS_ENABLED(CONFIG_BT_CTLR_DF)
+	err = lll_df_init();
+	if (err) {
+		return err;
+	}
+#endif
+
```
Move to line 406 after BT_CONN and before USER_EXT

ppryga replied:
Moved to requested line.

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_df.c:
```diff
@@ -0,0 +1,72 @@
+{
+	int err;
+
+	err = radio_df_ant_configure();
```
if the configuration will not be changed after power up, then place `radio_df_ant_configure` only in `lll_df_init`

ppryga replied:
At the moment it will be not changed after power up.
I did following change:
- removed `init_reset()` because it is no longer needed,
- `radio_df_ant_configure` is called directly in `lll_df_init`
- added `lll_df_reset` to `perform_lll_reset`.

cvinayak replied:
Please keep the `init_reset()` as a template for future implementation, the compiler removes it if its empty. Otherwise, changes in this PR looks good to me.

ppryga replied:
`init_reset()` restored

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_df.c:
```diff
@@ -0,0 +1,70 @@
+	return 0;
+}
+
+static int init_reset(void)
+{
+	return 0;
+}
+
```
Minor nitpick... move this to after non file static functions (to the end of this file).

ppryga replied:
function moved to bottom of the file

ppryga commented:
@carlescufi @anangl @ioannisg guys, could you do a review for the PR?

ppryga commented:
@cvinayak can you check the PR? Last change, move of the function `init_reset` to bottom of the file was implemented.

ppryga commented:
Rebased to resolve merge conflicts.

@ioannisg @anangl @carlescufi could you do a review?

ppryga commented:
@mbolivar-nordic @pabigot could you take a look on the PR?

pabigot commented on dts/bindings/bluetooth/nordic,nrf-dfe-ant.yaml:
```diff
@@ -0,0 +1,60 @@
+include: [base.yaml]
+
+properties:
+    ant-num:
```
AFAICT this property is unnecessary, as it can be calculated by the number of `dfegpio#-pin` properties that are valid.  I can't say offhand how to do that at compiletime, but I'm sure it's possible.  Otherwise just modify `radio_df_ant_num_get` to compute it from the pin array, maybe caching the value in a local `static uint8_t` if you expect to need it frequently.

Either way seems a better approach than having a property that might be inconsistent with the other properties.

anangl replied:
Perhaps values returned by `DT_NODE_HAS_PROP()` for all `dfegpioX-pin` properties could be just summed up?

ppryga replied:
We can't get information about available antenna number from dfegpio#-pin e.g. to drive Nordic proprietary antenna matrix with 12 antennae you have to use 4 GPIO pins. 
The property is added to handle such situations.

pabigot replied:
OK, that makes more sense.  A naive reader would assume each pin selects or deselects a single antenna.

pabigot reviewed (commented):
Some suggestions on the binding.  I don't feel qualified to approve or reject this.

anangl commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio_df.c:
```diff
@@ -0,0 +1,126 @@
+
+	for (uint8_t idx = 0; idx < DF_PSEL_GPIO_NUM; ++idx) {
+		if (ant_cfg.dfe_gpio[idx] != DF_GPIO_PIN_NOT_SET) {
+			NRF_RADIO->PSEL.DFEGPIO[idx] = ant_cfg.dfe_gpio[idx];
```
Use `nrf_radio_dfe_pattern_pin_set()` for this, please.

ppryga replied:
Changed as suggested.

anangl commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio_df.c:
```diff
@@ -0,0 +1,126 @@
+struct df_ant_cfg {
+	uint8_t ant_num;
+	/* Maximum supported antenna switch pattern length */
+	uint8_t max_ant_sw_pattern_num;
```
This field does not seem to be used in any way.

ppryga replied:
I will delete it.

ppryga replied:
corrected

ppryga commented on dts/bindings/bluetooth/nordic,nrf-dfe-ant.yaml:
```diff
@@ -0,0 +1,60 @@
+#  depends on end user which Radio PSEL_GPIO[n] registers will be used to drive
+#  antenna switches.
+#  Nevertheless there is a requirement that at least two antennae are available.
+#  That means at least one dfegpio[n]-pin must be provided, to give Radio
```
With single GPIO pin you are able to switch between two antennas (0 - ant#1, 1 - ant#2). 
GPIOs here are used to drive switches that are responsible for directing RF signal from radio peripheral to particular antenna.

pabigot replied:
Can you add that to the description somehow?  Or document (if true) that these provide alternatives to ant#1 which is always available?

ppryga replied:
OK, I'll try to document it more clearly.

BTW. It it not related with antenna available on Nordics DKs. To antenna matrix RF signal is bypassed from Radio peripheral to another antenna by SMA output. If bypass is not used, then Radio will use antenna on DK no matter of there configs and configuration done in Radio registers.

ppryga replied:
Added explanation for what are dfegpio#-pins used.

ppryga commented on dts/bindings/bluetooth/nordic,nrf-dfe-ant.yaml:
```diff
@@ -0,0 +1,60 @@
+#  possibility to switch between these two antennae.
+
+description: |
+    Nordic DFE antennae matrix description
```
Antennae is a plural form of "antenna" but I will update it if that is used in documentation.

ppryga commented on dts/arm/nordic/nrf52833.dtsi:
```diff
@@ -367,6 +367,12 @@
 			label = "WDT";
 		};
+
+		dfe_ant: dfe_ant {
```
I will add it there also when run the code on particular hardware. Currently we support only nRF52833 and due to that I've added it to this dtsi.

mbolivar-nordic replied:
Now that the compatible has changed to `nordic,nrf-radio`, I think it should be a "proper" node corresponding to the RADIO peripheral, like so:

```
radio: radio@40001000 {
        compatible = "nordic,nrf-radio"
        reg = <0x40001000 0x1000>;
        interrupts = <1 NRF_DEFAULT_IRQ_PRIORITY>;
        status = "okay";
};
```

I would appreciate a double-check on those values from @cvinayak or another domain expert.

The status here is "okay" because I think we want to indicate that it is enabled by default. This is distinct from CONFIG_BT, which can still be set to `n` to disable software support for the radio. Otherwise it will be too much hassle to add overlays enabling the radio to every single BT sample.

cvinayak replied:
Controller has not been using DTS, but it is desired to move towards when time permits (contributions always welcome). I have no objections to having radio node.

anangl commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio_df.c:
```diff
@@ -0,0 +1,126 @@
+
+#define DFE_ANT DT_NODELABEL(dfe_ant)
+#define DFE_GPIO_PIN(idx)						\
+	UTIL_OR(UTIL_AND(DT_NODE_HAS_PROP(DFE_ANT, dfegpio##idx##_pin),	\
+				DT_PROP(DFE_ANT, dfegpio##idx##_pin)),	\
+		DF_GPIO_PIN_NOT_SET)
```
This won't work for pin P0.00. Is that one not supposed to be ever used by this mechanism?

ppryga replied:
Why it won't work? 
According to UTIL_AND documentation:

> Like **a && b**, but does evaluation and
> short-circuiting at C preprocessor time.
> 
> This is not the same as the binary &&, however; in particular,
> **a** should expand to an integer literal 0 or 1. However, **b**
> can be any value.

That means DT_PROP(DFE_ANT, dfegpio##idx##_pin) is allowed to be zero.
If there is dfe_gpio#_pin = <0>, zero should be returned by complete macro.

anangl replied:
But then you'll have `UTIL_OR(0, DF_GPIO_PIN_NOT_SET)`, right?

ppryga replied:
You are right. I should use here COND_CODE_1 macro instead

ppryga replied:
corrected

anangl commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio_df.c:
```diff
@@ -0,0 +1,126 @@
+#error "DF antenna switching feature requires dfe_ant to be enabled in DTS"
+#endif
+
+#endif /* CONFIG_BT_CTLR_DF_ANT_SWITCH_TX ||
+	* CONFIG_BT_CTLR_DF_ANT_SWITCH_RX
+	*/
```
Can't this fit in one line?

ppryga replied:
It fits, will correct it.

ppryga replied:
corrected

anangl commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio_df.c:
```diff
@@ -0,0 +1,126 @@
+	 * for given antennae number. Confirm also that antennae number is not
+	 * lower than required minimum.
+	 */
+	if (((0x1 << pins_num) - 1) < ant_cfg.ant_num ||
```
Maybe using `BIT_MASK(pins_num)` here would make it a bit clearer.

ppryga replied:
will change

ppryga replied:
code removed due to: https://github.com/zephyrproject-rtos/zephyr/pull/30262/commits/44bfa9ba4f31e5cacddccd1fb2969c30a1dabc6b#r545135430

anangl commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio_df.c:
```diff
@@ -0,0 +1,126 @@
+	for (uint8_t idx = 0; idx < DF_PSEL_GPIO_NUM; ++idx) {
+		if (ant_cfg.dfe_gpio[idx] != DF_GPIO_PIN_NOT_SET) {
+			NRF_RADIO->PSEL.DFEGPIO[idx] = ant_cfg.dfe_gpio[idx];
+			++pins_num;
```
Why is it increased here?

ppryga replied:
It is a mistake. I will remove it.

ppryga replied:
Code removed.

anangl commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio_df.c:
```diff
@@ -0,0 +1,126 @@
+
+	for (uint8_t idx = 0; idx < DF_PSEL_GPIO_NUM; ++idx) {
+		if (ant_cfg.dfe_gpio[idx] != DF_GPIO_PIN_NOT_SET) {
+			++pins_num;
```
It seems that you could calculate this number at compile time (using `DT_NODE_HAS_PROP()`) and use a build assertion to check that `ant-num` is not set too high (as well as if it is not set too low).

ppryga replied:
changed to `BUILD_ASSERT`

mbolivar-nordic commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_df.c:
```diff
@@ -0,0 +1,70 @@
+ */
+int lll_df_reset(void)
+{
+	int err;
+
+	err = init_reset();
+	if (err) {
+		return err;
+	}
+
+	return 0;
```
```suggestion
	return init_reset();
```

ppryga replied:
Corrected

mbolivar-nordic commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_df.c:
```diff
@@ -0,0 +1,70 @@
+#include "hal/radio_df.h"
+#include "lll_df.h"
+
+/* @brief Function performs common steps for initialization and reset
```
These doxygen-style comments are using `/*` instead of `/**`. Is that intentional?

ppryga replied:
This is intentional, because there is no doxygen documentation generated for controller internal code. I wanted to add information about the function but I didn't want to generate documentation for that. 
Nevertheless I went to use doxygen markups to:
- have similar style of comments as in other header files
- if we decide to provide documentation for internals then simple change is required.

pabigot replied:
It doesn't hurt to make these `/**`, though I find that confusing for internal API as it normally suggests to me it's meant to be public.

I too use doxygen markup (e.g. `<at>param`) for internal functions in standard comment blocks, but I don't use `<at>brief` because it's annoying (and unnecessary with [JAVADOC_AUTOBRIEF](https://www.doxygen.nl/manual/config.html#cfg_javadoc_autobrief) which Zephyr didn't enable), so maybe people haven't noticed....

ppryga replied:
I think that we have to decide about common policy for internal API documentation.
In my humble opinion, it is good to have it, especially in cases where functions does more than name of the function suggest.

I leave it as is now. If it is requested to change, I'll do it throughout all my code in separate PR.

Should I create a github issue about internal API documentation format?

mbolivar-nordic commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_df.c:
```diff
@@ -0,0 +1,70 @@
+	return radio_df_ant_num_get();
+}
+
+static int init_reset(void)
+{
+	return 0;
+}
```
Shouldn't this be doing something?

ppryga replied:
At the moment it is not doing anything.
It is left on @cvinayak request - to use the same modules template. Code will be removed during compilation.

ppryga replied:
This is similar to above: https://github.com/zephyrproject-rtos/zephyr/pull/30262#discussion_r545686646

mbolivar-nordic commented on dts/bindings/bluetooth/nordic,nrf-dfe-ant.yaml:
```diff
@@ -0,0 +1,60 @@
+description: |
+    Nordic DFE antennae matrix description
+
+compatible: "nordic,nrf-dfe-ant"
```
These properties seem to be going straight into RADIO peripheral registers.

I might be wrong, but it really looks like we should be making a `nordic,nrf-radio` compatible instead.

pabigot replied:
That might be better, yes.

ppryga replied:
Shouldn't it go to radio directory if name of binding is changed to `nordic,nrf-radio`?

pabigot replied:
It should go into `dts/bindings/net/wireless` since we now have that from #30788.  That directory is where Linux puts radio hardware.

ppryga replied:
Moved to suggested location.

mbolivar-nordic commented on dts/bindings/bluetooth/nordic,nrf-dfe-ant.yaml:
```diff
@@ -0,0 +1,60 @@
+        Value zero means no antenna available. If provided value different than
+        zero, it must be greater or equal to two.
+
+    dfegpio0-pin:
+      type: int
+      required: false
+      description: Pin select for DFE pin 0
```
We are trying to get away from `foo-pin` properties; see https://github.com/zephyrproject-rtos/zephyr/pull/30769. I would really rather not see any more of these added upstream; we're just going to have to deprecate and remove them.

@anangl , @pabigot : what would you say to modifying #30769 so we don't have to resolve the I2C question, merging that in, and rebasing this PR using a `dfe-gpios` phandle-array instead?

pabigot replied:
Is there anything left to resolve on the I2C one?  0 is fine.

To switch to a phandle array: what would be used as the specifier for registers that aren't used?  There's no requirement that the enabled/configured GPIO settings be contiguous in the 8-register PSEL area.

mbolivar-nordic replied:
> Is there anything left to resolve on the I2C one? 0 is fine.

Yes, I believe so: https://github.com/zephyrproject-rtos/zephyr/pull/30769#issuecomment-747579873

I think this means we should support other flags in there, but TBH I'm not quite clear on the details and we might need a lot of discussion to get consensus, so I was thinking of decoupling that.



> To switch to a phandle array: what would be used as the specifier for registers that aren't used? There's no requirement that the enabled/configured GPIO settings be contiguous in the 8-register PSEL area.

2 options:

1. `df0-gpios`, `df1-gpios`, .... instead
2. an unspecified value in a phandle array can be given by setting the phandle to 0 and not giving any cells, like this: https://github.com/torvalds/linux/blob/0c6c887835b59c10602add88057c9c06f265effe/include/linux/of_gpio.h#L82

Option 2 doesn't work right now due to an edtlib bug. I'm looking into a fix.

pabigot replied:
The register that this sets is `PSEL.DFEGPIO[n]`.  I think the name should reflect that so `dfegpio#` seems right to me.  `-pin` if this continues to use the ordinal; if we go with #30769 then `-gpios`.

Option 2 is probably the devicetree way, but I suspect it'd be pretty confusing

mbolivar-nordic replied:
> Option 2 doesn't work right now due to an edtlib bug. I'm looking into a fix.

https://github.com/zephyrproject-rtos/zephyr/pull/30847

ppryga replied:
I've got three questions here: 
1. If I use -gpios instead of -pin, should I implement it as single property that is array of phandles or leave it as a separate properties for  each register as it is now ?
2. What about GPIO initialization sequence if the same GPIO is used for something else? Radio acquires those GPIOs only for IQ sampling during TX/RX. Besides that GPIOs are free to be used for something else. I do not say that it is a real world use case (someone would ask for that) but it is allowed by SOC spec. 
3. How can I translate port in -gpios property to a number? It is needed because value written to PSEL.DFEGPIO[n] have Nordic "encoding" of port and pin.

ppryga commented on dts/bindings/bluetooth/nordic,nrf-dfe-ant.yaml:
```diff
@@ -0,0 +1,60 @@
+# Copyright (c) 2020 Nordic Semiconductor ASA
+# SPDX-License-Identifier: Apache-2.0
+#
+# Note:
```
Moved to description property of the file.

ppryga commented on dts/bindings/bluetooth/nordic,nrf-dfe-ant.yaml:
```diff
@@ -0,0 +1,60 @@
+    ant-num:
+      type: int
+      required: true
+      description: >
```
Changed as requested.

ppryga commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio_df.c:
```diff
@@ -0,0 +1,126 @@
+
+/* @brief Function provides number of available antennae for Direction Finding.
+ *
+ * The number of antennae is hardware defined. It is provided via device tree.
```
I've corrected it.

anangl commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio_df.c:
```diff
@@ -0,0 +1,116 @@
+#define DFE_GPIO_PIN_DISCONNECT (RADIO_PSEL_DFEGPIO_CONNECT_Disconnected << \
+				 RADIO_PSEL_DFEGPIO_CONNECT_Pos)
+
+#define DFE_GPIO_PIN_CHECK(idx, _) COND_CODE_1(                         \
+	DT_NODE_HAS_PROP(DFE_ANT, dfegpio##idx##_pin), (1), (EMPTY)),
+
+#define DFE_GPIO_PIN_LIST LIST_DROP_EMPTY(                              \
+	UTIL_LISTIFY(DF_PSEL_GPIO_NUM, DFE_GPIO_PIN_CHECK, _))
+
+#define DFE_GPIO_NUM (NUM_VA_ARGS_LESS_1(DFE_GPIO_PIN_LIST) + 1)
```
This is wrong. It will give you 1 instead of 0 when no pins are defined. You need to use something like this:
```
#define COUNT_GPIO(idx, _)  + DT_NODE_HAS_PROP(DFE_ANT, dfegpio##idx##_pin)
#define DFE_GPIO_NUM        UTIL_LISTIFY(DF_PSEL_GPIO_NUM, COUNT_GPIO, _)
```

ppryga replied:
Thank you, I've corrected it. 
I have had to add additional verification if `DFE_GPIO_NUM > 0` to correctly use it in BUID_ASSERT.

anangl commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio_df.c:
```diff
@@ -0,0 +1,116 @@
+ */
+BUILD_ASSERT((BIT_MASK(DFE_GPIO_NUM) >= DT_PROP(DFE_ANT, ant_num)),     \
+	     "Insufficient number of GPIO pins configured.");
+BUILD_ASSERT((DT_PROP(DFE_ANT, ant_num) > DF_ANT_NUM_MIN),              \
```
`>=`

ppryga replied:
Thank you! I've updated it.

ppryga-nordic committed:
Bluetooth: controller: Fix error when max ant. pattern len not defined

Fix compilation error if CONFIG_BT_CTLR_DF is not defined.
If DF was not enabled, there was missing CONFIG_BT_CTLR_DF_MAX_-
ANT_SW_PATTERN_LEN and undeclared symbol was found.

Split ll_df.h into two parts:
- ll_df.h that is generally includeable and allows to successfully
  compile sources when DF is not enabled.
- ll_df_internal.h to be used in soucres that are compiled when
  DF is enabled.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga commented:
Fixed build error in unit tests.

anangl commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio_df.c:
```diff
@@ -0,0 +1,122 @@
+#define DFE_GPIO_PIN_DISCONNECT (RADIO_PSEL_DFEGPIO_CONNECT_Disconnected << \
+				 RADIO_PSEL_DFEGPIO_CONNECT_Pos)
+
+#define DFE_GPIO_PIN_CHECK(idx, _) COND_CODE_1(                         \
+	DT_NODE_HAS_PROP(DFE_ANT, dfegpio##idx##_pin), (1), (EMPTY)),
```
This seems to be no longer needed.

ppryga commented:
Removed unused macro and added comment to explain usage of DFE_GPIO_NUM_IS_ZERO macro.
@mbolivar-nordic can you do a review of the current state?

anangl commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio_df.c:
```diff
@@ -0,0 +1,122 @@
+#define DFE_GPIO_PIN_DISCONNECT (RADIO_PSEL_DFEGPIO_CONNECT_Disconnected << \
+				 RADIO_PSEL_DFEGPIO_CONNECT_Pos)
+
+#define DFE_GPIO_NUM        (0 UTIL_LISTIFY(DF_PSEL_GPIO_NUM, COUNT_GPIO, _))
```
`COUNT_GPIO` is not defined now. You need to restore one line removed in the last update.

ppryga replied:
Done, thank you very much for spotting that

ppryga commented:
Hello all!

I've implemented a change related with Marti's PR: https://github.com/zephyrproject-rtos/zephyr/pull/30769#event-4214075937.
`dfegpio#-pin` are removed and `dfegpio#-gpios` are used instead.

`dfegpio#-gpios` flag cell is expected to be set to zero. If it is not code will not compile. 
I've added `BUILD_ASSERT` to check that ->`DFE_GPIO_PIN_IS_FLAG_ZERO`.
I've done that because `dfegpio#-gpios` properties are not used to do any GPIO pins initialization.
To avoid possible confusion about it `BUILD_ASSERT` was added.

@mbolivar, @anangl, @pabigot please do another review. I'd like to finally merge the PR. Thank you in advance.

ppryga commented:
Fixe part of the compliance issues. The one that have left seems to me that are false-positives.

pabigot commented on dts/bindings/net/wireless/nordic,nrf-radio.yaml:
```diff
@@ -0,0 +1,103 @@
+        Value zero means no antenna available. If provided value different than
+        zero, it must be greater or equal to two.
+
+    dfegpio0-gpios:
```
I think somewhere there was discussion about whether things like this should be instead `dfegpio-gpios`, with the value being a sequence of phandle+specifiers.  IIRC that was at one point prevented by the inability to express an absent entry, but I think that was to be fixed, e.g.:
```
dfe-gpios = <&gpio0 0 0>, 0, <&gpio0 2 0>;
```

I have no position on this, but the array form would eliminate some verbosity.

ppryga replied:
I've decided to go with `dfegpios#-gpios`, because it gives end user possibility to use e.g. GPIO 1, 2, 5, 6.
Also I wanted to avoid use of programming macros in first attempt. It changed during review of @anangl.

pabigot commented on scripts/ci/check_compliance.py:
```diff
@@ -213,7 +213,12 @@ def run(self):
         except subprocess.CalledProcessError as ex:
             output = ex.output.decode("utf-8")
-            self.add_failure(output)
+            if re.search("[1-9][0-9]* errors,", output):
```
This doesn't belong in this PR.  It should be in a separate PR, whether or not there's an issue describing the problem.

ppryga replied:
I've removed this commit and will create a separate PR for that.

pabigot reviewed (commented):
One comment on devicetree style, not required.

The commit to check_compliance needs to go in a different PR, not be hidden in this one.

ppryga commented:
I've fixed two checkpatch warnings (related with line continuation). The one left is false-positive and according to @nashif https://github.com/zephyrproject-rtos/zephyr/pull/31413#issuecomment-762507556 should be marked/fixed by someone reponsible for merging. I'm sorry, I do not know the process. Whom should I ask for it to make the PR mergeable?

ppryga commented:
@pabigot @anangl @mbolivar-nordic @cvinayak gentlemen can you take a look. I need to merge the PR until end of the week.

pabigot reviewed (commented):
I have no objections to this, but still don't feel qualified to approve its technical components.

mbolivar-nordic commented on dts/bindings/net/wireless/nordic,nrf-radio.yaml:
```diff
@@ -0,0 +1,103 @@
+# SPDX-License-Identifier: Apache-2.0
+
+description: |
+    Nordic DFE antennas matrix description
```
```suggestion
    Nordic nRF family RADIO.
```

To match the other nRF bindings, since you've changed the compatible (and thanks for that).

ppryga replied:
Corrected, thanks for spotting that.

mbolivar-nordic commented on dts/bindings/net/wireless/nordic,nrf-radio.yaml:
```diff
@@ -0,0 +1,103 @@
+description: |
+    Nordic DFE antennas matrix description
+
+    dfegpio[n]-gpios properties provide GPIO pins to Radio peripheral. Provided
```
```suggestion
    dfegpio[n]-gpios properties provide GPIO pins to the radio peripheral. These
```

ppryga replied:
Changed according to suggestion

mbolivar-nordic commented on dts/bindings/net/wireless/nordic,nrf-radio.yaml:
```diff
@@ -0,0 +1,103 @@
+
+    dfegpio[n]-gpios properties provide GPIO pins to Radio peripheral. Provided
+    GPIO pins are used to drive antennas switching by Radio Direction Finding
+    Extension. E.g. to use 12 antennas 4 GPIO pins must be provided,
```
```suggestion
    Extension. E.g. to use 12 antennas 4 GPIO pins must be provided;
```

ppryga replied:
Changed according to suggestion

mbolivar-nordic commented on dts/bindings/net/wireless/nordic,nrf-radio.yaml:
```diff
@@ -0,0 +1,103 @@
+    Extension. E.g. to use 12 antennas 4 GPIO pins must be provided,
+    4 GPIO pins support switching up to 16 antennas.
+
+    All dfegpio[n]-gpios properties have required property set to false.
+    It depends on end user which Radio PSEL_GPIO[n] registers will be used
+    to drive antenna switches.
+    Nevertheless there is a requirement that at least two antennas are
+    available. That means at least one dfegpio[n]-gpios property must be
+    provided, to give Radio possibility to switch between these two antennas.
```
```suggestion
    The dfegpio[n] properties are used to initialize the PSEL.DFEGPIO[n] registers.
    None of the dfegpio[n]-gpios properties are marked required, since
    which Radio PSEL.DFEGPIO[n] registers will be used to drive antenna switches
    depends on the use case.
    
    Nevertheless, at least two antennas must be available. That means at
    least one dfegpio[n]-gpios property must be provided, to give the radio
    the possibility to switch between two antennas.
```

This is just some suggested rewording. I'm also suggesting using the register names in the product specification. I am not trying to change the meaning of your description.

ppryga replied:
Changed according to suggestion

mbolivar-nordic commented on dts/bindings/net/wireless/nordic,nrf-radio.yaml:
```diff
@@ -0,0 +1,103 @@
+include: [base.yaml]
+
+properties:
+    ant-num:
```
I think this should be renamed to `dfe-ant-num`, to indicate it's just for the number of switchable antennas for the direction finding extension.

ppryga replied:
I've changed the dfe_ant to proposed radio node. I've verified interrupt, register address and size. They are correct.
Also ant-num is now dfe-ant-num.

mbolivar-nordic reviewed (commented):
Just a few minor things. Thanks very much for the update.

ppryga commented:
I've implemented changes proposed by @mbolivar-nordic.
I've also reworded description in commit: https://github.com/zephyrproject-rtos/zephyr/pull/30262/commits/5fe418c65183037f7e7cc81d9c5d3a8df082bab7

cvinayak reviewed (approved):
My changes have been addressed in the controller subsystem.

mbolivar-nordic reviewed (approved):
The binding looks good to me now. Thanks for your patience and perseverance; nice work.
